### ISSUE TITLE
Array bounds are not checked in jupyter

### ISSUE THREAD

**Issue by sbliven** [SacBase/sac-jupyter#5]
The jupyter kernel appears not to check array bounds at runtime. This allows buffer overflows to expose memory segments.

Example:

```
x = [1][1];
```

should raise an error like

```
error: argument #1 of "_sel_VxA_" should be legal index into argument #2; types
error: found: int[1]{1}  and  int[1]{1}
```

However, in jupyter I get

```
Dimension:  0
Shape    : < >
 0
```

I can see that this is dumping memory by using it to initialize an array:

```
reshape([100],[])
```
```
Dimension:  1
Shape    : < 100>
< 0  0  0  0  1  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 -1 -1  0  0 520192  0 -146381216 1564  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 > 
```

**Comment by sbliven** [comment]
BTW, I noticed this because Exercise 5 doesn't give an error for `reshape([3,0,5], [])[[1,0]]` as I had expected.

**Comment by sbliven** [comment]
This might be a more general problem of not detecting runtime errors. For instance, dimensional mismatch is not detected either:

```
[[1]]+[1]
```
In jupyter gives
```
Dimension:  1
Shape    : <  1>
< 2 > 
```
Not
```
error: Shape of argument #1 of "_sel_VxA_" should match dimensionality of
error: argument #2; types found: int[1]  and  int[1,1]
```

**Comment by sbscholz** [comment]
you can activate the checks.

if you enter `%flags`, you will see the currently set flags:

```
-v0 -O0 -noprelude -noinl -specmode aud
```
which in essence disables all optimisations and does not include any checks.

One can change this manually using `%setflags <new-flags>`, e.g., 
```
%setflags -v0 -O0 -noprelude -noinl -specmode aud -check tc
```

Having said that...... I just tried it. It turns out that our checking for the AUD case is faulty :-(
So in order to get that error checked, one currently needs to enable specialisation to AKD, e.g. by using

```
%setflags -v0 -O0 -noprelude -noinl -specmode akd -check tc
```

**Comment by sbliven** [comment]
I tried using AKD but it causes some errors for valid code, e.g.

```
reshape([], [1])
*** SAC runtime error
*** Assignment with incompatible types found!
```

```
shift([0, -22], 0, mat)
*** SAC runtime error
*** Non-negativity constraint violated for scalar SACp_emal_4945__uprf_3517 in /tmp/jup-sac5yx8d6um/tmp0tfljh0r.exe.c:1742.
```

I don't know enough about AKD specialization to know whether these errors are expected, but it seems like the jupyter default flags should accept the same language as the sac2c default flags. (I'm using glot.io to compare, if that makes a difference.)

**Comment by sbscholz** [comment]
wow, amazing!
Goes to show that our testing is primarily done with the standard options enabled......

Here are several things at play.

**case 1: [[1]]+[1]   and -check tc.**    The reason why this does not flag up the error ultimately lies in the suppression of  all inlining (`-noinl`) in combination with the module system. Currently, we compile all modules without any checks. Since we inline all of the core functions, compilation with `-check tc` normally works fine, even if we use functions from our modules.
If, however, we compile with `-noinl`, then our only chance to get runtime checks for functions from modules happens, when we specialise functions as this leads to new compilations.... if we suppress specialisations (`-specmode aud`), then we end up calling the precompiled functions which do not contain any checks; even if we compile with `-check tc`.

The only proper solution to this problem (that I can see :-) is to make the compilation with `-check` its own SBI (SaC Binary Interface). That way, we would have different versions of modules around, one with checks and one without checks. The main reason this has not been done yet is that we already have quite a few SBIs around (multi-threading / cuda / mpi / ....) and we would need to now create a cross-product of them all .....

**case 2: reshape([], [1])** Here we seem to hit a bug in `-check t` which seems to be triggered by `-noinl`. Further investigation is needed here.

**case 3:shift([0, -22], 0, mat)** This is a bug in the constraint generation.... and a nasty one. What we have here boils down to a with-loop with an empty generator where the upper bound is negative. A smaller example is 
```
   a = with {
          ( [0] <= iv < [-5]) : 42;
       } : genarray( [12], 0);
```
By the time `-check c` was implemented this was illegal code. Later, we made this legal through a transformation in phase 9 (ewl). Since the check are injected in phase 8 (popt), this is flagged as an error despite the transformation in phase 9 modifying the upper bound to a positive number..... see https://gitlab.sac-home.org/sac-group/sac2c/-/issues/2312.
